### ISSUE TITLE
Pyright gets confused when a class that follows the singleton pattern has a decorator that calls `inspect.isclass()`

### ISSUE THREAD

**Issue by frank-lenormand** [microsoft/pyright#5289]
**Describe the bug**
A class that has a `classmethod` that returns an instance of that class (singleton pattern) makes Pyright produce an error (false-positive) .

I could only reproduce the issue when the decorator, which returns `cls` as-is, and also invokes `inspect.isclass(cls)` beforehand. If the call to `isclass` is removed, the error does not occur.

Reading the reproducer should help understand better what I mean.

**Reproducer**

```py3
#!/usr/bin/env python3


def test():
    def decorator(cls):
        # Removing these two lines or changing them to anything else makes Pyright skip
        import inspect
        assert inspect.isclass(cls)
        setattr(cls, "print_42", lambda _: print("pwnt"))
        return cls
    return decorator


@test()
class Test:
    _instance = None

    @classmethod
    def get(cls):
        if not cls._instance:
            cls._instance = cls()
        return cls._instance

    def print_42(self):
        print("42")


# “pwnt”
Test().print_42()

# “pwnt”
# Pyright: (only when the decorator calls `inspect.isclass(cls)`)
# test.py:32:6 - error: Cannot access member "get" for type "type"
#   Member "get" is unknown (reportGeneralTypeIssues)
Test.get().print_42()
```

**Expected behavior**
Pyright does not produce any errors.

**VS Code extension or command-line**

Pyright: `v1.1.314`

Command-line.

Reproduced with a `pipx` installation and `npm` in a Docker container.

**Comment by erictraut** [comment]
This behavior is expected, so I don't consider it a bug.

Your decorator is unannotated, so pyright is applying return type inference to determine its type. If you remove the `assert` in the decorator code, the return type is `Unknown`. But if you include the `assert`, its type becomes `type` (because the `inspect.isclass` is defined with as `def isclass(object: object) -> TypeGuard[type[Any]]: ...`.

To fix this error, you need to annotate your decorator as follows:

```python
_T = TypeVar("_T")

def test() -> Callable[[type[_T]], type[_T]]:
    def decorator(cls: type[_T]) -> type[_T]:
```

**Comment by frank-lenormand** [comment]
Thank you!
